pkg/datastore/mysql/model_wrapper.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+func wrapModel(entity interface{}) (interface{}, error) {
+	switch e := entity.(type) {
+	case *model.Project:
+		if e == nil {
+			return nil, fmt.Errorf("nil entity given")
+		}
```

Comment:
Instead of checking kind name like [here](https://github.com/pipe-cd/pipe/pull/1653/files#diff-c47f267fba561fafc1be077abac3f330be6605efdd9290475d39cfe1a055fef0R21), how about making reproducible uuid from Project.ID using hash. For instance:

```go
	space, err := uuid.FromBytes([]byte("aaaaaaaaaaaaaaaa"))
	if err != nil {
		return
	}
	i := uuid.NewSHA1(space, []byte("aaaaaaaaaaaaaaaaaaaaaaaaaa"))
```

Reply:
I know generating it each time is high cost. But operating against Project is rarely created/udpated relatively. If we adopt this solution, I feel like we can remove the tedious check for kind == "Project". But what do you think?

Reply:
I see your point 👍 We will completely ignore the `id` passed as Update/Create function's parameter and use `model.Id` instead for all models, am I guess right? In case we keep project.ID as currently, that's okay to adopt your suggestion, but I wonder if we plan to change Project.ID from simple string to UUID and make name attribute as other models, then this could be unnecessary. How do you think? 👀 
cc @nghialv

Reply:
@nakabonne Sorry, maybe I misunderstand your suggestion, could you clarify it? 👀 
Are we going to make UUID from string in case entity of type `model.Project` and save it as `Id` in new create wrapped `mysql.project` type?

Reply:
>Are we going to make UUID from string in case entity of type model.Project and save it as Id in new create wrapped mysql.project type?

Right, right. That's what I was trying to say!

I'm thinking about kind of like:

```go
		// Defined in advance
		space, err := uuid.FromBytes([]byte("aaaaaaaaaaaaaaaa"))
		if err != nil {
			return
		}
		e.Id = uuid.NewSHA1(space, []byte(e.Id)).String()
```

>I wonder if we plan to change Project.ID from simple string to UUID and make name attribute as other models

But if it's gonna be as you mentioned, it's absolutely unnecessary.

Reply:
But eventually even for this solution, we have to convert the project id into the UUID.

```go
// makeUUID converts the given id into a reproducible UUID.
func makeUUID(id []byte) string {
	return uuid.NewSHA1(space, []byte(id)).String()
}

func (m *MySQL) Get(ctx context.Context, kind, id string, v interface{}) error {
	if kind == "Project" { // eventually this check needed
		id = makeUUID([]byte(id))
	}
...
```

In that sense, I feel like this solution isn't the best one as well. As you said, it would be nice if we could change the Project.ID to UUID from the current one (such as "pipecd")

Reply:
Thanks for clarify 🙏 
Then there is another thing concerned me 👀 If we "encode" projectId to uuid type and set it as `id` attribute of json data, then we have to decode it on returns (query for `data` field from table.) Currently I'm trying to avoid change the JSON data as much as possible to avoid encode/decode and keep consistent on data value. 😁

Reply:
Exactly, that is so true. The solution I suggested has a couple of downsides especially in getting data. Okay, please never mind!
Thank you for giving me the opportunity to think about handling JSON in the SQL context!

Reply:
I have a neutral view on this ID hashing because the ID inside the JSON field should be kept as the original value.

One possible way is storing the hashed value in the ID column of the Project table. (no change on this wrapper)
By that way, we can remove the kind checking inside `buildGetQuery` and `buildUpdateQuery` functions.
But that way also introduces the kind checking and calling of the following function at other places such as inside `Get` and `Update` functions.

``` golang
func hashProjectID(id string) string {
    return uuid.NewSHA1(uuid.NameSpaceOID, []byte(id)).String()
}
```

Reply:
> the ID inside the JSON field should be kept as the original value.

Yep,  I have the same opinion on this 👍 Let's keep it as is for now and come back later if necessary 😁

Reply:
>Let's keep it as is for now and come back later if necessary

👍 

>One possible way is storing the hashed value in the ID column of the Project table.

Thank you. This is exactly what I was thinking. But eventually we have to check the kind is "Project" so that it converts the project id given by clients into the UUID.

```go
func (m *MySQL) Get(ctx context.Context, kind, id string, v interface{}) error {
	if kind == "Project" { // eventually this check needed
		id = hashProjectID(id)
	}
...
```

So it seems reasonable to conclude that the ID hashing solution is not good so much.